flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/GateBuffersSpec.java

```diff
@@ -0,0 +1,122 @@
+ * Streaming jobs is {#link NettyShuffleUtils#DEFAULT_MAX_BUFFERS_PER_GATE_FOR_STREAMING}.
+ */
+public class GateBuffersSpec {
+
+    private final int effectiveExclusiveBuffersPerChannel;
+
+    private final int minOptionalBuffers;
+
+    private final int maxOptionalBuffers;
+
+    private final int maxEffectiveTotalBuffersPerGate;
+
+    private GateBuffersSpec(
+            int effectiveExclusiveBuffersPerChannel,
+            int minOptionalBuffers,
+            int maxOptionalBuffers,
+            int maxEffectiveTotalBuffersPerGate) {
+        this.effectiveExclusiveBuffersPerChannel = effectiveExclusiveBuffersPerChannel;
+        this.minOptionalBuffers = minOptionalBuffers;
+        this.maxOptionalBuffers = maxOptionalBuffers;
+        this.maxEffectiveTotalBuffersPerGate = maxEffectiveTotalBuffersPerGate;
+    }
+
+    int minOptionalBuffers() {
+        return minOptionalBuffers;
+    }
+
+    int maxOptionalBuffers() {
+        return maxOptionalBuffers;
+    }
+
+    int getEffectiveExclusiveBuffersPerChannel() {
+        return effectiveExclusiveBuffersPerChannel;
+    }
+
+    public int maxEffectiveTotalGateBuffers() {
+        return maxEffectiveTotalBuffersPerGate;
+    }
+
+    public static GateBuffersSpec from(
+            Optional<Integer> configuredMaxRequiredBuffersPerGate,
+            int configuredNetworkBuffersPerChannel,
+            int configuredFloatingNetworkBuffersPerGate,
+            ResultPartitionType partitionType,
+            int numInputChannels) {
+        int maxRequiredBuffersThresholdPerGate =
+                getEffectiveMaxRequiredBuffersPerGate(
+                        partitionType, configuredMaxRequiredBuffersPerGate);
+        int minBuffersTargetPerGate =
+                getMinBuffersTargetPerGate(numInputChannels, configuredNetworkBuffersPerChannel);
+        int maxBuffersTargetPerGate =
+                getMaxBuffersTargetPerGate(
+                        numInputChannels,
+                        configuredNetworkBuffersPerChannel,
+                        configuredFloatingNetworkBuffersPerGate);
+        int minRequiredBuffersPerGate =
+                Math.min(maxRequiredBuffersThresholdPerGate, minBuffersTargetPerGate);
+
+        int effectiveExclusiveBuffersPerChannel =
+                adjustExclusiveBuffersPerChannel(
+                        configuredNetworkBuffersPerChannel,
+                        numInputChannels,
+                        minRequiredBuffersPerGate);
+        int effectiveExclusiveBuffersPerGate =
+                getEffectiveExclusiveBuffersPerGate(
+                        numInputChannels, effectiveExclusiveBuffersPerChannel);
+
+        int minOptionalBuffers = minRequiredBuffersPerGate - effectiveExclusiveBuffersPerGate;
+        int maxOptionalBuffers = maxBuffersTargetPerGate - effectiveExclusiveBuffersPerGate;
+        int maxEffectiveTotalBuffersPerGate = effectiveExclusiveBuffersPerGate + maxOptionalBuffers;
```

Comment:
Shouldn't this always be `maxBuffersTargetPerGate`?

Reply:
Yes, I have updated it to return `maxBuffersTargetPerGate `.